### ISSUE TITLE
syscall/js: Make syscall/js optional when compiling wasm?

### ISSUE THREAD

**Issue by justinclift** [golang/go#27766]
At present, when we compile Wasm `syscall/js` is automatically included.

This forces the runtime to either be a browser (eg Firefox, Chrome, etc), or at least pretend to be one.

That's kind of non-optimal, as some of the wasm execution environments presently in-development aren't targeted at browser environments. eg:

* https://github.com/Xe/olin
* https://github.com/ewasm/hera

In the [WebAssembly](https://gophers.slack.com/messages/C4GU6CHSQ) Gophers Slack channel we have people asking about non-browser use cases fairly often.  It seems likely that'll be a fairly standard use case, if that can be catered to as well.

How feasible would it be to have some way to suppress the default `syscall/js` inclusion, or otherwise make it optional?

**Comment by Xe** [comment]
I have investigated [creating a custom GOOS in the past](https://github.com/Xe/go/commit/37bc30ec064a32a630cdad325f83020c1f5be976), but it fails with that attached gnarly linker error. Guidance on it would be helpful, as implementing this would unblock the use of Go for my server side webassembly things.

**Comment by sbinet** [comment]
Loosely related to https://github.com/golang/go/issues/25612 and github.com/go-interpreter/wagon

**Comment by peterbourgon** [comment]
Little to add except that I also have use cases for Go-to-wasm that don't involve a browser, and I'd be happy to chat about them to anyone interested in the issue.

**Comment by bcmills** [comment]
CC @neelance @cherrymui

**Comment by neelance** [comment]
If you don't have `syscall/js`, then you'd need something else to interact with the environment. Since there currently is no standard API in the WebAssembly ecosystem, I don't see why we should invent yet another API on our own. If such a standard API emerges in the future, I'd be happy to adopt it.

Today, I would implement the backend API of `syscall/js` as much as needed for the standard libraries, but no further. You don't need to implement full JavaScript for it. Do you see any significant downsides to this approach?

**Comment by justinclift** [comment]
@Xe @peterbourgon Probably best for you to jump in here. :smile:

**Comment by neelance** [comment]
Is there any effort to get https://github.com/CommonWA/cwa-spec adopted by https://github.com/WebAssembly, maybe as an official spec?

**Comment by justinclift** [comment]
CC @losfair

**Comment by peterbourgon** [comment]
@neelance 

> If you don't have `syscall/js`, then you'd need something else to interact with the environment. 

Essentially I want to punt on this question. I want to compile Go down to a WebAssembly module, exporting a set of functions as a sort of public API, and binding to or declaring (somehow) a set of import functions, to consume other modules. I want to load and run this module in a server-side environment, with no relation to JS. CommonWA is not of much interest to me (us) in this context.

**Comment by neelance** [comment]
@peterbourgon 

Would you want to call the functions concurrently or only one at a time?
Would they do HTTP requests to other services?

**Comment by Xe** [comment]
@peterbourgon in this case using CommonWA (note: I am not married to CommonWA, it's just not something made by me and looks like it could easily be the basis we could build off of as a community to make it better for everyone) would us all just be standardizing on the set of things the binaries the Go compiler emits would be depending on.

**Comment by justinclift** [comment]
@peterbourgon 
> If you're looking for some kind of examples, I can try to dig something up; let me know.

Please do, they'd likely help. :smile:

**Comment by DavidHuie** [comment]
I have a similar use case to @peterbourgon in mind. Wasm on the server creates the opportunity for new types of features, such as user developed, server executed extensions to products. I'd love to use Go for these sorts of things!

**Comment by neelance** [comment]
Sorry, but I still don't see why `syscall/js` prevents people from using non-JavaScript hosts. It is only an API, it should be possible to implement it without having full-blown JavaScript.

**Comment by Xe** [comment]
@neelance What is the minimal implementation required for every function to have to return an error-like response?

**Comment by neelance** [comment]
@Xe That of course depends on the function. I would just start with an empty stub of each function which logs its execution and then add the implementation step by step.

**Comment by Xe** [comment]
Yes, can you please provide these empty stubs? I can't figure them out on my own apparently.

**Comment by neelance** [comment]
I don't even know which WebAssembly host you are trying to use.

**Comment by neelance** [comment]
The import functions that you need to stub are the ones you find in
https://github.com/golang/go/blob/7a6ccece877232fc58872c85fdf0718629b83d91/misc/wasm/wasm_exec.js#L199

**Comment by Xe** [comment]
For all of these functions: https://github.com/golang/go/blob/master/src/syscall/js/js_js.s

- What is the stack layout in relative ordinal bytes?
- What values are arguments?
- What values are returns?
- What is the "zero" return?
- What is a C-like analogue for each of these functions?

**Comment by neelance** [comment]
It's the normal Go calling convention. All arguments and return values live on the stack. You can find the function signatures in https://github.com/golang/go/blob/master/src/html/template/js.go and in the wasm_exec.js I linked above. You can also find the stack offsets in the wasm_exec.js if you don't want to calculate them on your own. A quick Google search got me this: https://science.raphael.poss.name/go-calling-convention-x86-64.html

**Comment by neelance** [comment]
C has a different calling convention than Go, so you won't be able to use simple C function signatures.

**Comment by neelance** [comment]
@Xe Sorry, I don't have that right now. The `os` package will also require `syscall/js` to not return zero all the time. I can only recommend that you start with a simple example, add what's necessary to make it work and then continue from there. I can help you further if you have a specific error that you don't know how to resolve.

**Comment by Xe** [comment]
Then if it shouldn't return 0 all the time, what is the valid thing to do?

**Comment by neelance** [comment]
Fulfill the interface good enough that it supports the actions that you want to support or that it at least does not crash on initialization. Specifics depend on the case.

**Comment by Xe** [comment]
Can you please rephrase that in a way that tells me more than nothing?

**Comment by neelance** [comment]
> I can help you further if you have a specific error that you don't know how to resolve.

This still holds. I don't see how I can give you more information without implementing it myself. There is no specification that you can fulfill in one go. You'll have to start with some concrete example that you want to make pass and then we can figure out together what is necessary to do so.

**Comment by justinclift** [comment]
Thanks heaps @neelance. :smile:

**Comment by backkem** [comment]
Exploring options for supporting different types of IO in neelance/go_js_wasm_exec#1.

**Comment by backkem** [comment]
There is initial work being done on a CloudABI target at [NuxiNL/go](https://github.com/NuxiNL/go). This is the system ABI the [CraneStation/wasmtime](https://github.com/CraneStation/wasmtime) runtime plans to use.

**Comment by justinclift** [comment]
This is probably relevant as well: https://hacks.mozilla.org/2019/03/standardizing-wasi-a-webassembly-system-interface/

**Comment by neelance** [comment]
Awesome! This is exactly what I was waiting for. I'll look into it soon.

**Comment by tiborvass** [comment]
Does this mean we'll have a `GOOS=wasi` ?

**Comment by peterbourgon** [comment]
@neelance It was suggested to me that anyone working on wasi for Go might want to join #wasi on the [Mozilla IRC](https://wiki.mozilla.org/IRC)—just passing that info along.

**Comment by justinclift** [comment]
#31105 sounds related too. :smile:

**Comment by mfateev** [comment]
I believe for execution Go doesn't need any environment support. Even random seed and time can be hardcoded and it is fine for a lot of use cases. Forcing every host to provide a dummy implementation of WASI or syscall/js sounds very artificial. BTW when Rust and C compile to wasi they don't have any required imports.

Ideally I would love to have the compiler options to not include some Go standard library components at all. For example I would prefer that in some cases the code that references any file or socket IO wouldn't even compile.

**Comment by neelance** [comment]
> I believe for execution Go doesn't need any environment support.

If there is no environment, then your Go code can calculate something, but then only discard the result.

> when Rust and C compile to wasi they don't have any required imports.

Neither Rust nor C has a runtime/goroutines. You can directly call a Rust/C function from your WebAssembly host. This is not easily possible with Go's semantics.

**Comment by mastersingh24** [comment]
[TinyGo](https://tinygo.org/) does not have this issue as they do not assume a JS host environment.

**Comment by rsc** [comment]
This issue is titled "make syscall/js optional" while #31105 is instead to assume wasi.
As noted above, you have to have _something_ or else there's no way to interact with the outside system.
Perhaps this issue should be closed as a duplicate of #31105?

For wasi, https://github.com/golang/go/issues/31105#issuecomment-884696837 describes a way to use Go with WASI.

**Comment by justinclift** [comment]
> Perhaps this issue should be closed as a duplicate of #31105?

Not sure personally.  I don't use Go for wasm any longer, so no longer have any err... skin in this issue.

If other people do, I'll close or leave this open as seems appropriate. :smile:
